File: android/src/main/scala/com/yhsif/notifbot/HttpSender.scala
package com.yhsif.notifbot

import android.os.AsyncTask

import java.io.IOException

import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response

object HttpSender {
  val KeyLabel = "label"
  val KeyMsg = "msg"
  val CodeNetFail: Int = -1

  private val client =
    new OkHttpClient.Builder().followRedirects(false).build()

  def send(
      url: String,
      label: String,
      msg: String,
      onSuccess: () => Unit,
      onFailure: () => Unit,
      onNetFail: () => Unit): Unit = {
    val body =
      new FormBody.Builder().add(KeyLabel, label).add(KeyMsg, msg).build()
    val request = new Request.Builder().url(url).post(body).build()

    new HttpSender(onSuccess, onFailure, onNetFail).execute(request)
  }

  def checkUrl(url: String, onFailure: () => Unit): Unit = {
    val request = new Request.Builder().url(url).get().build()
    new HttpSender(() => {}, onFailure, () => {}).execute(request)
  }
}

class HttpSender(
    val onSuccess: () => Unit,
    val onFailure: () => Unit,
    val onNetFail: () => Unit)
    extends AsyncTask[AnyRef, AnyRef, AnyRef] {

  override def doInBackground(reqs: AnyRef*): AnyRef = {
    reqs.foreach { req =>
      // Only handle the first req
      try {
        val res = HttpSender.client.newCall(req.asInstanceOf[Request]).execute()
        if (res == null) {
          return HttpSender.CodeNetFail.asInstanceOf[AnyRef]
        }
        val code = res.code().asInstanceOf[AnyRef]
        res.close()
        return code
      } catch {
        case _: IOException => {
          return HttpSender.CodeNetFail.asInstanceOf[AnyRef]
        }
      }
    }
    // Empty reqs
    return 404.asInstanceOf[AnyRef]
  }

  override def onPostExecute(c: AnyRef): Unit = {
    val code = c.asInstanceOf[Int]
    if (code == HttpSender.CodeNetFail) {
      onNetFail()
      return
    }
    if (code >= 200 && code < 400) {
      onSuccess()
    } else {
      onFailure()
    }
  }
}

File: android/src/com/yhsif/notifbot/HttpSender.kt
package com.yhsif.notifbot

import android.content.Context
import android.net.Uri
import com.google.android.gms.net.CronetProviderInstaller
import com.google.android.gms.tasks.Tasks
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.asExecutor
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.chromium.net.CronetEngine
import org.chromium.net.CronetException
import org.chromium.net.UploadDataProviders
import org.chromium.net.UrlRequest
import org.chromium.net.UrlResponseInfo
import org.chromium.net.impl.JavaCronetProvider
import java.nio.ByteBuffer
import java.util.concurrent.ExecutionException

class HttpSender(
  val onSuccess: () -> Unit,
  val onFailure: () -> Unit,
  val onNetFail: () -> Unit,
) : UrlRequest.Callback() {

  companion object {
    private const val KEY_LABEL = "label"
    private const val KEY_MSG = "msg"

    private lateinit var context: Context

    private val engine: CronetEngine by lazy {
      lateinit var builder: CronetEngine.Builder
      try {
        Tasks.await(CronetProviderInstaller.installProvider(context))
        builder = CronetEngine.Builder(context)
      } catch (e: ExecutionException) {
        builder = JavaCronetProvider(context).createBuilder()
      }
      builder.enableHttp2(true).enableQuic(true).build()
    }

    fun initEngine(ctx: Context) {
      if (!::context.isInitialized) {
        context = ctx
      }
    }

    fun send(
      ctx: Context,
      url: String,
      label: String,
      msg: String,
      onSuccess: () -> Unit,
      onFailure: () -> Unit,
      onNetFail: () -> Unit,
    ) {
      initEngine(ctx)
      GlobalScope.launch(Dispatchers.IO) {
        val body = Uri.Builder()
          .appendQueryParameter(KEY_LABEL, label)
          .appendQueryParameter(KEY_MSG, msg)
          .build()
          .getEncodedQuery()!!

        val reqBuilder = engine.newUrlRequestBuilder(
          url,
          HttpSender(onSuccess, onFailure, onNetFail),
          Dispatchers.IO.asExecutor(),
        )
        reqBuilder.setHttpMethod("POST")
        reqBuilder.addHeader("Content-Type", "application/x-www-form-urlencoded")
        reqBuilder.setUploadDataProvider(
          UploadDataProviders.create(body.toByteArray(), 0, body.length),
          Dispatchers.IO.asExecutor(),
        )
        reqBuilder.build().start()
      }
    }

    fun checkUrl(url: String, onSuccess: () -> Unit, onFailure: () -> Unit) {
      GlobalScope.launch(Dispatchers.IO) {
        engine.newUrlRequestBuilder(
          url,
          HttpSender(
            onSuccess,
            onFailure,
            {},
          ),
          Dispatchers.IO.asExecutor(),
        ).build().start()
      }
    }
  }

  override fun onRedirectReceived(req: UrlRequest, info: UrlResponseInfo?, newUrl: String) {
    // Never follow redirects, but treat it as success
    req.cancel()
    GlobalScope.launch(Dispatchers.Main) {
      onSuccess()
    }
  }

  override fun onResponseStarted(req: UrlRequest, info: UrlResponseInfo?) {
    val code = info?.getHttpStatusCode()
    GlobalScope.launch(Dispatchers.Main) {
      if (code != null && code >= 200 && code < 400) {
        onSuccess()
      } else {
        onFailure()
      }
      withContext(Dispatchers.IO) {
        req.cancel()
      }
    }
  }

  override fun onFailed(req: UrlRequest, info: UrlResponseInfo?, e: CronetException) {
    GlobalScope.launch(Dispatchers.Main) {
      onNetFail()
    }
  }

  // We don't care about the following functions
  override fun onReadCompleted(req: UrlRequest, info: UrlResponseInfo?, buf: ByteBuffer) {}

  override fun onSucceeded(req: UrlRequest, info: UrlResponseInfo?) {}

  override fun onCanceled(req: UrlRequest, info: UrlResponseInfo?) {}
}
